File: packages/autonavi_search_flutter/ios/Classes/SearchChannelHandler.swift
import Flutter

class SearchChannelHandler: NSObject, FlutterPlugin {

    private let adapter = AMapSearchAdapter()

    override init() { super.init() }

    public static func register(with registrar: FlutterPluginRegistrar) {}

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        let args = call.arguments as? [String: Any] ?? [:]

        switch call.method {
        case "search#keyword":
            handleKeywordSearch(args, result: result)
        case "search#nearby":
            handleNearbySearch(args, result: result)
        case "search#regeocode":
            handleRegeocode(args, result: result)
        case "search#geocode":
            handleGeocode(args, result: result)
        case "route#driving":
            handleDrivingRoute(args, result: result)
        case "route#walking":
            handleWalkingRoute(args, result: result)
        case "search#district":
            handleDistrictSearch(args, result: result)
        default:
            result(FlutterMethodNotImplemented)
        }
    }

    // MARK: - Search Handlers

    private func handleKeywordSearch(_ args: [String: Any], result: @escaping FlutterResult) {
        adapter.searchKeyword(
            keyword:  args["keyword"]  as? String ?? "",
            city:     args["city"]     as? String ?? "",
            types:    args["types"]    as? String,
            pageSize: args["pageSize"] as? Int ?? 20,
            page:     args["page"]     as? Int ?? 1
        ) { page, err in
            if let err = err { result(FlutterError(code: err.code, message: err.message, details: nil)); return }
            result(poiPageToMap(page!))
        }
    }

    private func handleNearbySearch(_ args: [String: Any], result: @escaping FlutterResult) {
        adapter.searchNearby(
            latitude:  args["latitude"]  as? Double ?? 0,
            longitude: args["longitude"] as? Double ?? 0,
            radius:    args["radius"]    as? Int ?? 1000,
            keyword:   args["keyword"]   as? String,
            types:     args["types"]     as? String,
            pageSize:  args["pageSize"]  as? Int ?? 20,
            page:      args["page"]      as? Int ?? 1
        ) { page, err in
            if let err = err { result(FlutterError(code: err.code, message: err.message, details: nil)); return }
            result(poiPageToMap(page!))
        }
    }

    private func handleRegeocode(_ args: [String: Any], result: @escaping FlutterResult) {
        adapter.regeocode(
            latitude:  args["latitude"]  as? Double ?? 0,
            longitude: args["longitude"] as? Double ?? 0
        ) { ac, err in
            if let err = err { result(FlutterError(code: err.code, message: err.message, details: nil)); return }
            guard let ac = ac else { result(nil); return }
            result([
                "formattedAddress": ac.formattedAddress, "country": ac.country,
                "province": ac.province, "city": ac.city, "cityCode": ac.cityCode,
                "district": ac.district, "adCode": ac.adCode, "street": ac.street,
                "streetNumber": ac.streetNumber, "township": ac.township, "townCode": ac.townCode,
            ] as [String: Any?])
        }
    }

    private func handleGeocode(_ args: [String: Any], result: @escaping FlutterResult) {
        adapter.geocode(
            address: args["address"] as? String ?? "",
            city:    args["city"]    as? String
        ) { items, err in
            if let err = err { result(FlutterError(code: err.code, message: err.message, details: nil)); return }
            let maps = (items ?? []).map { gc -> [String: Any?] in
                ["formattedAddress": gc.formattedAddress, "country": gc.country,
                 "province": gc.province, "city": gc.city, "cityCode": gc.cityCode,
                 "district": gc.district, "adCode": gc.adCode,
                 "latitude": gc.latitude, "longitude": gc.longitude, "level": gc.level]
            }
            result(maps)
        }
    }

    private func handleDrivingRoute(_ args: [String: Any], result: @escaping FlutterResult) {
        guard let originMap = args["origin"] as? [String: Any],
              let destMap   = args["destination"] as? [String: Any] else {
            result(FlutterError(code: "INVALID_ARGS", message: "origin/destination required", details: nil))
            return
        }
        let waypoints = (args["waypoints"] as? [[String: Any]] ?? []).map {
            (lat: $0["latitude"] as? Double ?? 0, lng: $0["longitude"] as? Double ?? 0)
        }
        adapter.drivingRoute(
            originLat: originMap["latitude"] as? Double ?? 0,
            originLng: originMap["longitude"] as? Double ?? 0,
            destLat:   destMap["latitude"]   as? Double ?? 0,
            destLng:   destMap["longitude"]  as? Double ?? 0,
            waypoints: waypoints
        ) { paths, err in
            if let err = err { result(FlutterError(code: err.code, message: err.message, details: nil)); return }
            result(["paths": (paths ?? []).map { routePathToMap($0) }])
        }
    }

    private func handleWalkingRoute(_ args: [String: Any], result: @escaping FlutterResult) {
        guard let originMap = args["origin"] as? [String: Any],
              let destMap   = args["destination"] as? [String: Any] else {
            result(FlutterError(code: "INVALID_ARGS", message: "origin/destination required", details: nil))
            return
        }
        adapter.walkingRoute(
            originLat: originMap["latitude"] as? Double ?? 0,
            originLng: originMap["longitude"] as? Double ?? 0,
            destLat:   destMap["latitude"]   as? Double ?? 0,
            destLng:   destMap["longitude"]  as? Double ?? 0
        ) { paths, err in
            if let err = err { result(FlutterError(code: err.code, message: err.message, details: nil)); return }
            result(["paths": (paths ?? []).map { routePathToMap($0) }])
        }
    }

    private func handleDistrictSearch(_ args: [String: Any], result: @escaping FlutterResult) {
        adapter.searchDistrict(
            keywords: args["keywords"] as? String ?? "",
            level:    args["level"]    as? Int ?? 3
        ) { items, err in
            if let err = err { result(FlutterError(code: err.code, message: err.message, details: nil)); return }
            result((items ?? []).map { districtToMap($0) })
        }
    }
}

// MARK: - Map conversion helpers (free functions, no SDK imports)

private func poiPageToMap(_ page: AMapSearchAdapter.POIPage) -> [String: Any] {
    return [
        "pois": page.pois.map { poi -> [String: Any?] in
            ["poiId": poi.poiId, "title": poi.title, "typeDes": poi.typeDes,
             "typeCode": poi.typeCode, "latitude": poi.latitude, "longitude": poi.longitude,
             "address": poi.address, "tel": poi.tel, "distance": poi.distance,
             "cityName": poi.cityName, "adName": poi.adName, "snippet": poi.address]
        },
        "totalCount": page.totalCount,
        "pageCount":  page.pageCount,
        "pageNum":    page.pageNum,
    ]
}

private func routePathToMap(_ path: AMapSearchAdapter.RoutePath) -> [String: Any] {
    return [
        "distance":     path.distance,
        "duration":     path.duration,
        "strategy":     path.strategy,
        "tolls":        path.tolls,
        "tollDistance": path.tollDistance,
        "trafficLights":path.trafficLights,
        "steps": path.steps.map { step -> [String: Any?] in
            ["instruction": step.instruction, "road": step.road,
             "distance": step.distance, "duration": step.duration,
             "action": step.action, "path": step.path]
        },
    ]
}

private func districtToMap(_ d: AMapSearchAdapter.DistrictItem) -> [String: Any?] {
    return [
        "name": d.name, "adCode": d.adCode, "cityCode": d.cityCode, "level": d.level,
        "latitude": d.latitude, "longitude": d.longitude,
        "districts": d.districts.map { districtToMap($0) },
    ]
}

File: packages/autonavi_search_flutter/android/src/main/kotlin/com/autonavi/search/flutter/SearchChannelHandler.kt
package com.autonavi.search.flutter

import android.content.Context
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel

class SearchChannelHandler(private val context: Context) : MethodChannel.MethodCallHandler {

    private val adapter = AMapSearchAdapter(context)

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        @Suppress("UNCHECKED_CAST")
        val args = call.arguments as? Map<String, Any> ?: emptyMap()

        when (call.method) {
            "search#keyword"  -> searchKeyword(args, result)
            "search#nearby"   -> searchNearby(args, result)
            "search#regeocode"-> regeocode(args, result)
            "search#geocode"  -> geocode(args, result)
            "route#driving"   -> drivingRoute(args, result)
            "route#walking"   -> walkingRoute(args, result)
            "search#district" -> searchDistrict(args, result)
            else -> result.notImplemented()
        }
    }

    private fun searchKeyword(args: Map<String, Any>, result: MethodChannel.Result) {
        adapter.searchKeyword(
            keyword  = args["keyword"]  as String,
            city     = args["city"]     as String,
            types    = args["types"]    as? String ?: "",
            page     = (args["page"]     as? Int) ?: 1,
            pageSize = (args["pageSize"] as? Int) ?: 20,
        ) { page, err ->
            if (err != null) { result.error(err.first, err.second, null); return@searchKeyword }
            result.success(poiPageToMap(page!!))
        }
    }

    private fun searchNearby(args: Map<String, Any>, result: MethodChannel.Result) {
        adapter.searchNearby(
            lat      = (args["latitude"]  as Number).toDouble(),
            lng      = (args["longitude"] as Number).toDouble(),
            radius   = (args["radius"]    as? Int) ?: 1000,
            keyword  = args["keyword"]    as? String ?: "",
            types    = args["types"]      as? String ?: "",
            page     = (args["page"]      as? Int) ?: 1,
            pageSize = (args["pageSize"]  as? Int) ?: 20,
        ) { page, err ->
            if (err != null) { result.error(err.first, err.second, null); return@searchNearby }
            result.success(poiPageToMap(page!!))
        }
    }

    private fun regeocode(args: Map<String, Any>, result: MethodChannel.Result) {
        adapter.regeocode(
            lat = (args["latitude"]  as Number).toDouble(),
            lng = (args["longitude"] as Number).toDouble(),
        ) { ac, err ->
            if (err != null) { result.error(err.first, err.second, null); return@regeocode }
            result.success(ac?.let {
                mapOf(
                    "formattedAddress" to it.formattedAddress,
                    "country"   to it.country,   "province"  to it.province,
                    "city"      to it.city,       "cityCode"  to it.cityCode,
                    "district"  to it.district,   "adCode"    to it.adCode,
                    "street"    to it.street,      "streetNumber" to it.streetNumber,
                    "township"  to it.township,   "townCode"  to it.townCode,
                )
            })
        }
    }

    private fun geocode(args: Map<String, Any>, result: MethodChannel.Result) {
        adapter.geocode(
            address = args["address"] as String,
            city    = args["city"]    as? String ?: "",
        ) { items, err ->
            if (err != null) { result.error(err.first, err.second, null); return@geocode }
            result.success((items ?: emptyList()).map { gc ->
                mapOf(
                    "formattedAddress" to gc.formattedAddress,
                    "country"   to gc.country,  "province" to gc.province,
                    "city"      to gc.city,      "district" to gc.district,
                    "adCode"    to gc.adCode,
                    "latitude"  to gc.latitude,  "longitude" to gc.longitude,
                    "level"     to gc.level,
                )
            })
        }
    }

    @Suppress("UNCHECKED_CAST")
    private fun drivingRoute(args: Map<String, Any>, result: MethodChannel.Result) {
        val originMap = args["origin"]      as Map<String, Any>
        val destMap   = args["destination"] as Map<String, Any>
        val waypoints = (args["waypoints"] as? List<Map<String, Any>>)?.map {
            (it["latitude"] as Number).toDouble() to (it["longitude"] as Number).toDouble()
        } ?: emptyList()

        adapter.drivingRoute(
            originLat = (originMap["latitude"]  as Number).toDouble(),
            originLng = (originMap["longitude"] as Number).toDouble(),
            destLat   = (destMap["latitude"]    as Number).toDouble(),
            destLng   = (destMap["longitude"]   as Number).toDouble(),
            waypoints = waypoints,
        ) { paths, err ->
            if (err != null) { result.error(err.first, err.second, null); return@drivingRoute }
            result.success(mapOf("paths" to (paths ?: emptyList()).map { routePathToMap(it) }))
        }
    }

    @Suppress("UNCHECKED_CAST")
    private fun walkingRoute(args: Map<String, Any>, result: MethodChannel.Result) {
        val originMap = args["origin"]      as Map<String, Any>
        val destMap   = args["destination"] as Map<String, Any>

        adapter.walkingRoute(
            originLat = (originMap["latitude"]  as Number).toDouble(),
            originLng = (originMap["longitude"] as Number).toDouble(),
            destLat   = (destMap["latitude"]    as Number).toDouble(),
            destLng   = (destMap["longitude"]   as Number).toDouble(),
        ) { paths, err ->
            if (err != null) { result.error(err.first, err.second, null); return@walkingRoute }
            result.success(mapOf("paths" to (paths ?: emptyList()).map { routePathToMap(it) }))
        }
    }

    private fun searchDistrict(args: Map<String, Any>, result: MethodChannel.Result) {
        adapter.searchDistrict(keywords = args["keywords"] as String) { items, err ->
            if (err != null) { result.error(err.first, err.second, null); return@searchDistrict }
            result.success((items ?: emptyList()).map { districtToMap(it) })
        }
    }

    // MARK: - Map conversion helpers (no SDK types)

    private fun poiPageToMap(page: AMapSearchAdapter.POIPage): Map<String, Any> = mapOf(
        "pois" to page.pois.map { poi ->
            mapOf(
                "poiId"    to poi.poiId,    "title"   to poi.title,
                "typeDes"  to poi.typeDes,  "typeCode" to poi.typeCode,
                "latitude" to poi.latitude, "longitude" to poi.longitude,
                "address"  to poi.address,  "tel"     to poi.tel,
                "distance" to poi.distance, "cityName" to poi.cityName,
                "adName"   to poi.adName,   "snippet"  to poi.address,
            )
        },
        "totalCount" to page.totalCount,
        "pageCount"  to page.pageCount,
        "pageNum"    to page.pageNum,
    )

    private fun routePathToMap(path: AMapSearchAdapter.RoutePath): Map<String, Any?> = mapOf(
        "distance"     to path.distance,
        "duration"     to path.duration,
        "strategy"     to path.strategy,
        "tolls"        to path.tolls,
        "tollDistance" to path.tollDistance,
        "trafficLights" to path.trafficLights,
        "steps" to path.steps.map { step ->
            mapOf(
                "instruction" to step.instruction, "road"     to step.road,
                "distance"    to step.distance,    "duration" to step.duration,
                "action"      to step.action,       "path"    to step.path,
            )
        },
    )

    private fun districtToMap(d: AMapSearchAdapter.DistrictNode): Map<String, Any?> = mapOf(
        "name"      to d.name,     "adCode"   to d.adCode,
        "cityCode"  to d.cityCode, "level"    to d.level,
        "latitude"  to d.latitude, "longitude" to d.longitude,
        "districts" to d.districts.map { districtToMap(it) },
    )
}
